applications/crossbar/src/modules/cb_fmc_devices.erl

```diff
@@ -0,0 +1,285 @@
+-spec validate(cb_context:context()) -> cb_context:context().
+-spec validate(cb_context:context(), path_token()) -> cb_context:context().
+validate(Context) ->
+    validate_fmc_devices(Context, cb_context:req_verb(Context)).
+validate(Context, Id) ->
+    validate_fmc_device(Context, Id, cb_context:req_verb(Context)).
+
+-spec validate_fmc_devices(cb_context:context(), http_method()) -> cb_context:context().
+validate_fmc_devices(Context, ?HTTP_GET) ->
+    read(Context);
+validate_fmc_devices(Context, ?HTTP_PUT) ->
+    OnSuccess = fun(C) -> on_successful_validation('undefined', C) end,
+    cb_context:validate_request_data(<<"fmc_device">>, Context, OnSuccess).
+
+-spec validate_fmc_device(cb_context:context(), path_token(), http_method()) -> cb_context:context().
+validate_fmc_device(Context, Id, ?HTTP_GET) ->
+    case couch_mgr:open_cache_doc(?WH_FMC_DB, Id) of
+        {'ok', Doc} ->
+            AccountId = cb_context:account_id(Context),
+            case wh_json:get_value(<<"account_id">>, Doc) of
+                AccountId ->
+                    Context1 = cb_context:set_doc(Context, Doc),
+                    read(Context1, Id);
+                _ ->
+                    cb_context:add_validation_error(<<"fmc">>, <<"not_found">>, wh_json:from_list([{<<"message">>, <<"FMC Device with this ID wasn't found">>}]), Context)
+            end;
+        _ ->
+            cb_context:add_validation_error(<<"fmc">>, <<"not_found">>, wh_json:from_list([{<<"message">>, <<"FMC Device with this ID wasn't found">>}]), Context)
+    end;
+validate_fmc_device(Context, Id, ?HTTP_POST) ->
+    OnSuccess = fun(C) -> on_successful_validation(Id, C) end,
+    cb_context:validate_request_data(<<"fmc_device">>, Context, OnSuccess);
+validate_fmc_device(Context, Id, ?HTTP_DELETE) ->
+    case couch_mgr:open_cache_doc(?WH_FMC_DB, Id) of
+        {'ok', SavedDoc} ->
+            AccountId = cb_context:account_id(Context),
+            case wh_json:get_value(<<"account_id">>, SavedDoc) of
+                AccountId ->
+                    Context1 = cb_context:set_doc(Context, SavedDoc),
+                    delete(Context1, Id);
+                _ ->
+                    cb_context:add_validation_error(<<"fmc">>, <<"not_found">>, wh_json:from_list([{<<"message">>, <<"FMC Device with this ID wasn't found">>}]), Context)
+            end;
+        _ ->
+            cb_context:add_validation_error(<<"fmc">>, <<"not_found">>, wh_json:from_list([{<<"message">>, <<"FMC Device with this ID wasn't found">>}]), Context)
+    end.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Load an instance from the database
+%% @end
+%%--------------------------------------------------------------------
+-spec read(cb_context:context()) -> cb_context:context().
+read(Context) ->
+    {'ok', JObjs} = couch_mgr:get_all_results(?WH_FMC_DB, ?CB_LIST),
+    Doc = [wh_json:get_value(<<"value">>, JObj) || JObj <- JObjs],
```

Comment:
I'd like these list comprehensions merged to avoid traversing the list more than once.